Sources/IntegerUtilities/Shift.swift

```diff
@@ -0,0 +1,149 @@
+      // the desired rounding mode.
+      let count = count - Count(bitWidth - 1)
+      var floor = self >> count
+      let frac = self - (floor << count)
+      if frac != 0 { floor |= 1 } // insert sticky bit
+      return floor.shifted(right: bitWidth - 1, rounding: rule)
+    }
+    // Now we are in the happy case: 0 < count < bitWidth, which makes all
+    // the math to handle rounding simpler.
+    //
+    // TODO: If we were really, really careful about overflow, some of these
+    // could be made simpler. E.g. mathematically round up is implemented here
+    // via:
+    //
+    //   floor = self >> count
+    //   mask = 1 << count - 1
+    //   frac = self & mask
+    //   return floor + (frac == 0 ? 0 : 1)
+    //
+    // _if_ we didn't have to worry about intermediate overflow (either because
+    // self is bounded away from .max or because we don't have a fixed-width
+    // type), then we could use the following instead:
+    //
+    //   mask = 1 << count - 1
+    //   return (self + mask) >> count
+    //
+    // However, self + mask can overflow, e.g. if we're shifting .max by 1:
+    //
+    //   mask = 1 << 1 - 1 = 1
+    //   self + mask = .max + 1 // uh-oh
+    //
+    // This cannot occur for arbitrary-precision numbers, so we could use
+    // the simpler expressions for those (but those are precisely the cases
+    // where performance does not matter). More interesting, we could promote
+    // fixed-width numbers to a wider type, preventing this intermediate
+    // overflow and allowing us to use the simpler expressions much of the
+    // time. We could also explicitly _detect_ the overflow if it happens and
+    // patch up the result, though this is a little bit tricky because
+    // addingReportingOverflow does not exist on BinaryInteger. Hence, this
+    // is a TODO for the future.
+    let mask = Magnitude(1) << count - 1
+    let frac = Magnitude(truncatingIfNeeded: self) & mask
+    let floor = self >> count
+    let ceiling = floor + (frac == 0 ? 0 : 1)
+    switch rule {
+    case .down:
+      return floor
+    case .up:
+      return ceiling
+    case .towardZero:
+      return self > 0 ? floor : ceiling
+    case .awayFromZero:
+      return self < 0 ? floor : ceiling
+    case .toOdd:
+      return floor | (frac == 0 ? 0 : 1)
+    case .toNearestOrAwayFromZero:
+      let round = mask >> 1 + (self > 0 ? 1 : 0)
+      return floor + Self((round + frac) >> count)
+    case .toNearestOrEven:
+      let round = mask >> 1 + Magnitude(floor & 1)
+      return floor + Self((round + frac) >> count)
+    case .stochastic:
+      // TODO: it's unfortunate that we can't specify a custom random source
+      // for the stochastic rounding rule, but I don't see a nice way to have
+      // that share the API with the other rounding rules, because we'd then
+      // have to take the RNG in-out. The same problem applies to rounding
+      // with dithering. We should consider adding a stateful rounding API
```

Comment:
Is mentioning dithering relevant? I find it confusing, especially since I thought that it is exactly this case.

Reply:
The `//` comments are mostly notes for myself, rather than public documentation. =)

Reply:
I'm not sure I'd include stochastic rounding in this API.  It raises too many questions.  (Ulichney's "Digital Halftoning" is a great introduction that discusses the importance of controlling the frequency distribution of your noise source.)

Reply:
Blue noise and ordered dithers require a stateful rounding API, but even stochastic rounding without state is quite useful.